src/queries/query-evaluation-model-in-detail.md

```diff
@@ -0,0 +1,236 @@
+ - The only parameters a provider function takes are the key and a reference to
+   the "query context" (which provides access to rest of the "database").
+
+The database is built up lazily by invoking queries. The query providers will
+invoke other queries, for which the result is either already cached or computed
+by calling another query provider. These query provider invocations
+conceptually form a directed acyclic graph (DAG) at the leaves of which are
+input values that are already known when the query context is created.
+
+
+
+## Caching/Memoization
+
+Results of query invocations are "memoized" which means that the query context
+will cache the result in an internal table and, when the query is invoked with
+the same query key again, will return the result from the cache instead of
+running the provider again.
+
+This caching is crucial for making the query engine efficient. Without
+memoization the system would still be sound (that is, it would yield the same
+results) but the same computations would be done over and over again.
+
+Memoization is one of the main reasons why query providers have to be pure
+functions. If calling a provider function could yield different results for
+each invocation (because it accesses some global mutable state) then we could
+not memoize the result.
+
+
+
+## Input data
+
+When the query context is created, it is still empty: No queries have been
+executed, no results are cached. But the context already provides access to
+"input" data, i.e. pieces of immutable data that where computed before the
+context was created and that queries can access to do their computations.
+Currently this input data consists mainly of the HIR map and the command-line
+options the compiler was invoked with. In the future, inputs will just consist
+of command-line options and a list of source files -- the HIR map will itself
+be provided by a query which processes these source files.
+
+Without inputs, queries would live in a void without anything to compute their
+result from (remember, query providers only have access to other queries and
+the context but not any other outside state or information).
+
+For a query provider, input data and results of other queries look exactly the
+same: It just tells the context "give me the value of X". Because input data
+is immutable, the provider can rely on it being the same across
+different query invocations, just as is the case for query results.
+
+
+
+## An example execution trace of some queries
+
+How does this DAG of query invocations come into existence? At some point
+the compiler driver will create the, as yet empty, query context. It will then,
+from outside of the query system, invoke the queries it needs to perform its
+task. This looks something like the following:
+
+```rust,ignore
+fn compile_crate() {}
+    let cli_options = ...;
+    let hir_map = ...;
+
+    // Create the query context `tcx`
+    let tcx = TyCtxt::new(cli_options, hir_map);
```

Comment:
This is probably a dumb question, but why is the query context called `TyCtxt`? I'd have thought it would be called `QryCtxt` or `QCtxt` or something.

It's been that way as long as I can remember and I assumed it was short for "type context" or something.

Reply:
The `tcx` is called "type context" or "type checking context" purely for historical reasons, I'd say. There's been talk about renaming to "query context" (e.g. `qx` or `qcx`) for a while. For now, the `tcx` acts the query context (and does a few other things too).